File: solutions/c/perfect-numbers/1/perfect_numbers.c
#include "perfect_numbers.h"
#include <math.h>

kind classify_number(int number) {
    // Validate input
    if (number <= 0) {
        return ERROR;
    }
    
    // Special case: 1 has no proper divisors other than itself
    if (number == 1) {
        return DEFICIENT_NUMBER;
    }
    
    int aliquot_sum = 1; // 1 is always a divisor (except for number == 1)
    
    // Only need to check up to sqrt(number)
    int limit = (int)sqrt(number);
    
    for (int i = 2; i <= limit; i++) {
        if (number % i == 0) {
            // Add divisor i
            aliquot_sum += i;
            
            // Add the paired divisor (unless it's the same as i, i.e., square root)
            if (i != number / i) {
                aliquot_sum += number / i;
            }
        }
    }
    
    // Compare aliquot sum with original number
    if (aliquot_sum == number) {
        return PERFECT_NUMBER;
    } else if (aliquot_sum > number) {
        return ABUNDANT_NUMBER;
    } else {
        return DEFICIENT_NUMBER;
    }
}

File: solutions/fortran/perfect-numbers/1/perfect_numbers.f90
module perfect_numbers
  implicit none

contains

  character(len=9) function classify(num)
    integer, intent(in) :: num
    integer :: i, sum, limit

    ! Handle invalid input (non-positive numbers)
    if (num <= 0) then
      classify = "ERROR"
      return
    end if

    sum = 0
    limit = int(sqrt(real(num)))

    do i = 1, limit
      if (mod(num, i) == 0) then
        ! Add the divisor i if it is not the number itself
        if (i < num) sum = sum + i

        ! Add the complementary divisor if it is different from i and not the number itself
        if (i /= num / i .and. num / i < num) sum = sum + num / i
      end if
    end do

    if (sum == num) then
      classify = "perfect"
    else if (sum > num) then
      classify = "abundant"
    else
      classify = "deficient"
    end if
  end function classify

end module perfect_numbers